assets/app/crowbar-app.module.js

```diff
@@ -36,4 +37,27 @@
             $urlRouterProvider.otherwise('/upgrade/landing');
         }
     }
+
+    run.$inject = ['$rootScope', '$state', 'upgradeFactory', 'upgradeStepsFactory'];
+
+    function run($rootScope, $state, upgradeFactory, upgradeStepsFactory) {
```

Comment:
Implementing the main functionality in a service was exactly what I did here. I didn't want to imeplement `$on('$stateChangeStart')` handler in the factory though as this would require injecting many additional things there.
I thought I could make `lastStepForRestore()` function parameter-less but the status data is received in an async way so it would require some promise magic to make this work. This is why I did the `getStatus()` part on the app module side.
Note, that the main (and most complicated one) logic which requires testing is enclosed in `lastStepForRestore()` which is a pure function (except the localStorage part) and as such can be easily tested without any mocking etc.